evm/vm/evm.py

```diff
@@ -243,30 +236,42 @@ def import_block(self, block):
         except BlockNotFound:
             raise ValidationError("Parent ({0}) of block {1} not found".format(
                 block.header.parent_hash, block.header.hash))
+
         init_header = self.create_header_from_parent(parent_header)
         parent_evm = type(self)(self.db, init_header)
         imported_block = parent_evm.get_vm().import_block(block)
+        # It feels wrong to call validate_block() on self here, but we do that
+        # because we want to look up the recent uncles starting from the
+        # currenct canonical chain head.
         self.validate_block(imported_block)
 
         persist_block_to_db(self.db, imported_block)
-        # TODO: We must only do this when the imported block has higher
-        # score than current head.
-        self.header = self.create_header_from_parent(imported_block.header)
+        self.maybe_add_to_canonical_chain(imported_block)
 
         return imported_block
 
-    def mine_block(self, **mine_params):
-        """
-        Mine the current block, applying
-        """
-        block = self.get_vm().mine_block(**mine_params)
-        self.validate_block(block)
-        persist_block_to_db(self.db, block)
-
+    def maybe_add_to_canonical_chain(self, block):
+        current_head = self.get_block_by_hash(self.header.parent_hash)
+        if get_score(self.db, block.hash) <= get_score(self.db, current_head.hash):
+            return
+
+        new_chain = []
+        b = block
+        while b.number >= GENESIS_BLOCK_NUMBER:
+            new_chain.append(b)
```

Comment:
Oh, interesting, I'd seen this pattern of combining generators with the @to_tuple decorator and I wasn't sure what was the advantage, but now I understand you do it to avoid the mutation and make it more readable? Do you just find it's more readable than appending to a list and returning it at the end or is there anything else that makes you prefer this?